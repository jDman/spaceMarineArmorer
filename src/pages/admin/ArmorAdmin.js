import React, { Component } from 'react';
import axios from 'axios';

import Input from '../../components/form/input/Input';
import Button from '../../components/button/Button';
import PageTitle from '../../components/page-title/PageTitle';

import maxLengthCheck from '../../utils/maxLengthCheck';
import minLengthCheck from '../../utils/minLengthCheck';
import maxNumberCheck from '../../utils/maxNumberCheck';
import minNumberCheck from '../../utils/minNumberCheck';
import requiredCheck from '../../utils/requiredCheck';

import classes from './ArmorAdmin.module.css';

class ArmorAdmin extends Component {
  state = {
    loading: false,
    armorDetailsFormIsValid: false,
    armorDetailsForm: {
      name: {
        elementType: 'input',
        elementConfig: {
          type: 'text',
          name: 'name',
          required: true,
        },
        value: '',
        validators: [requiredCheck(null)],
        valid: false,
        touched: false,
      },
      type: {
        elementType: 'select',
        elementConfig: {
          selectAttributes: {
            type: 'select',
            name: 'type',
            required: true,
          },
          options: [
            { value: 'helmet', displayValue: 'Helmet' },
            { value: 'body', displayValue: 'Body' },
            { value: 'arm', displayValue: 'Arm' },
            { value: 'leg', displayValue: 'Leg' },
          ],
        },
        value: 'helmet',
        valid: true,
        validators: [],
        touched: false,
      },
      description: {
        elementType: 'textarea',
        elementConfig: {
          name: 'description',
          placeholder: 'This armor is...',
          required: true,
          minLength: 5,
          maxLength: 2000,
        },
        value: '',
        validators: [
          requiredCheck(null),
          minLengthCheck(5),
          maxLengthCheck(2000),
        ],
        valid: false,
        touched: false,
      },
      company: {
        elementType: 'select',
        elementConfig: {
          selectAttributes: {
            type: 'select',
            name: 'company',
            required: true,
          },
          options: [
            { value: 'starscape_systems', displayValue: 'Starscape Systems' },
            { value: 'adrax_corp', displayValue: 'Adrax Corp' },
            { value: 'orian_labs', displayValue: 'Orian Labs' },
          ],
        },
        value: 'starscape_systems',
        valid: true,
        validators: [],
        touched: false,
      },
      stock: {
        elementType: 'input',
        elementConfig: {
          type: 'number',
          name: 'stock',
          min: 0,
          max: 100,
          step: 1,
        },
        value: 0,
        valid: true,
        validators: [],
        touched: false,
      },
      discount: {
        elementType: 'input',
        elementConfig: {
          type: 'number',
          name: 'discount',
          min: 0,
          max: 100,
          step: 5,
        },
        value: 0,
        valid: true,
        validators: [],
        touched: false,
      },

      cost: {
        elementType: 'input',
        elementConfig: {
          type: 'number',
          name: 'cost',
          required: true,
          min: 0,
          max: 10000,
          step: 100,
        },
        value: 0,
        validators: [
          requiredCheck('0'),
          minNumberCheck(0),
          maxNumberCheck(10000),
        ],
        valid: false,
        touched: false,
      },
      shield: {
        elementType: 'input',
        elementConfig: {
          type: 'number',
          name: 'shield',
          min: 0,
          max: 100,
          step: 10,
        },
        value: 0,
        valid: true,
        validators: [],
        touched: false,
      },
      protection: {
        elementType: 'select',
        elementConfig: {
          selectAttributes: {
            type: 'select',
            name: 'protection',
            required: true,
          },
          options: [
            { value: 'high', displayValue: 'High' },
            { value: 'medium', displayValue: 'Medium' },
            { value: 'low', displayValue: 'Low' },
          ],
        },
        value: 'high',
        valid: true,
        validators: [],
        touched: false,
      },
      quality: {
        elementType: 'select',
        elementConfig: {
          selectAttributes: {
            type: 'select',
            name: 'quality',
            required: true,
          },
          options: [
            { value: 'high', displayValue: 'High' },
            { value: 'medium', displayValue: 'Medium' },
            { value: 'low', displayValue: 'Low' },
          ],
        },
        value: 'high',
        valid: true,
        validators: [],
        touched: false,
      },
    },
  };

  addArmor(event) {
    event.preventDefault();

    this.setState({ loading: true });

    const armorDetailsFormData = {};

    for (let key in this.state.armorDetailsForm) {
      armorDetailsFormData[key] = this.state.armorDetailsForm[key].value;
    }

    axios
      .post('http://localhost:4000/admin/armor/add', {
        ...armorDetailsFormData,
        createdBy: { userId: '5e70dfb438cee83fd9e004fd', userName: 'Freddy' },
      })
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  }

  inputChangedHandler(value, identifier) {
    this.setState((prevState) => {
      let isValid = true;

      for (const validator of prevState.armorDetailsForm[identifier]
        .validators) {
        isValid = isValid && validator(value);
      }

      const updatedForm = {
        ...prevState.armorDetailsForm,
        [identifier]: {
          ...prevState.armorDetailsForm[identifier],
          touched: true,
          valid: isValid,
          value: value,
        },
      };

      let formIsValid = true;

      for (const inputName in updatedForm) {
        formIsValid = formIsValid && updatedForm[inputName].valid;
      }

      return {
        armorDetailsForm: updatedForm,
        armorDetailsFormIsValid: formIsValid,
      };
    });
  }

  render() {
    const armorDetailsFormElements = [];
    const armorDetailsForm = this.state.armorDetailsForm;

    for (let key in armorDetailsForm) {
      armorDetailsFormElements.push({
        id: key,
        config: armorDetailsForm[key],
      });
    }

    return (
      <section className={classes.ArmorAdmin}>
        <PageTitle title="Add an Armor" />

        <form onSubmit={(event) => this.addArmor(event)}>
          {armorDetailsFormElements.map((formElement) => (
            <Input
              key={formElement.id}
              elementType={formElement.config.elementType}
              elementConfig={formElement.config.elementConfig}
              value={formElement.config.value}
              invalid={!formElement.config.valid}
              touched={formElement.config.touched}
              hasValidation={formElement.config.validators.length}
              changed={(event) =>
                this.inputChangedHandler(event.target.value, formElement.id)
              }
            />
          ))}
          <Button
            type="submit"
            classes={
              !this.state.armorDetailsFormIsValid
                ? [
                    classes.ArmorAdminAddArmorButton,
                    classes.ArmorAdminAddArmorButtonDisabled,
                  ]
                : classes.ArmorAdminAddArmorButton
            }
            disabled={!this.state.armorDetailsFormIsValid}
            buttonText="Add Armor"
          />
        </form>
      </section>
    );
  }
}

export default ArmorAdmin;
