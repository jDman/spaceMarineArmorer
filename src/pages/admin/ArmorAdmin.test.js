import React from 'react';
import renderer from 'react-test-renderer';
import ArmorAdmin from './ArmorAdmin';

describe('<ArmorAdmin />', () => {
  it('renders ArmorAdmin page component', () => {
    const component = renderer.create(<ArmorAdmin />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
