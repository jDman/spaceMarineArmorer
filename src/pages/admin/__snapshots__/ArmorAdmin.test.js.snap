// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArmorAdmin /> renders ArmorAdmin page component 1`] = `
<section>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="shield"
        onChange={[Function]}
        step={10}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="stock"
        onChange={[Function]}
        step={1}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="discount"
        onChange={[Function]}
        step={5}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="protection"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="helmet"
        >
          Helmet
        </option>
        <option
          value="body"
        >
          Body
        </option>
        <option
          value="arm"
        >
          Arm
        </option>
        <option
          value="leg"
        >
          Leg
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={10000}
        min={0}
        name="cost"
        onChange={[Function]}
        required={true}
        step={100}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="protection"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="high"
        >
          High
        </option>
        <option
          value="medium"
        >
          Medium
        </option>
        <option
          value="low"
        >
          Low
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="quality"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="high"
        >
          High
        </option>
        <option
          value="medium"
        >
          Medium
        </option>
        <option
          value="low"
        >
          Low
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <textarea
        className="InputElement"
        maxLength={2000}
        minLength={5}
        name="description"
        onChange={[Function]}
        placeholder="This armor is..."
        required={true}
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="company"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="starscape_systems"
        >
          Starscape Systems
        </option>
        <option
          value="adrax_corp"
        >
          Adrax Corp
        </option>
        <option
          value="orian_labs"
        >
          Orian Labs
        </option>
      </select>
    </div>
    <button
      disabled={true}
      type="submit"
    >
      Add
    </button>
  </form>
</section>
`;

exports[`renders ArmorAdmin page component 1`] = `
<section>
  <form
    onSubmit={[Function]}
  >
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="shield"
        onChange={[Function]}
        step={10}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="stock"
        onChange={[Function]}
        step={1}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={100}
        min={0}
        name="discount"
        onChange={[Function]}
        step={5}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="protection"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="helmet"
        >
          Helmet
        </option>
        <option
          value="body"
        >
          Body
        </option>
        <option
          value="arm"
        >
          Arm
        </option>
        <option
          value="leg"
        >
          Leg
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <input
        className="InputElement"
        max={10000}
        min={0}
        name="cost"
        onChange={[Function]}
        required={true}
        step={100}
        type="number"
        value={0}
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="protection"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="high"
        >
          High
        </option>
        <option
          value="medium"
        >
          Medium
        </option>
        <option
          value="low"
        >
          Low
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="quality"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="high"
        >
          High
        </option>
        <option
          value="medium"
        >
          Medium
        </option>
        <option
          value="low"
        >
          Low
        </option>
      </select>
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <textarea
        className="InputElement"
        maxLength={2000}
        minLength={5}
        name="description"
        onChange={[Function]}
        placeholder="This armor is..."
        required={true}
        value=""
      />
    </div>
    <div
      className="Input"
    >
      <label
        className="Label"
      />
      <select
        className="InputElement"
        name="company"
        onChange={[Function]}
        required={true}
        type="select"
        value=""
      >
        <option
          value="starscape_systems"
        >
          Starscape Systems
        </option>
        <option
          value="adrax_corp"
        >
          Adrax Corp
        </option>
        <option
          value="orian_labs"
        >
          Orian Labs
        </option>
      </select>
    </div>
    <button
      disabled={true}
      type="submit"
    >
      Add
    </button>
  </form>
</section>
`;
